### PR TITLE
main() wrapper

### PR DESCRIPTION
**Pull request by res2k**
## Description
This adds a wrapper around `main()` functions to deal with unhandled exceptions by printing a "fatal" log message and exiting.
The motivation is the behavior of unhandled exceptions on MSVC. There, such an exception simply terminates the program, but the exception message is not output anywhere! This can make troubleshooting very confusing, as the log of a tool gives no indication of a problem.

## Features list
Not really a feature.

## Implementation remarks
Wrapper is "opt-in" by including `system/main.hpp`.
The header defines a macro `main` so the `main()` function in the source appears like the real one even though it's wrapped. It's also convenient.

### CODE REVIEW

**Comment by fabiencastan** [comment]
Thanks for the contribution. I fully agree with the need!

Few remarks regarding the implementation:
1/ I would prefer to find a solution without having to create a define on "main". You can easily have a library with a member variable named "int main;" or something else.
We could just manually rename "main" by "aliceVision_main" in all main_*.cpp files so it is explicit and if the main.hpp is not included, the compilation will fail.

2/ I would keep main.hpp as a header only file and I would also remove the intermediate aliceVision_main_wrapper, as we can just put the try/catch in the real main.

What do think?

**Comment by res2k** [comment]
Hi,
I've made the changes you requested.

File: src/aliceVision/system/main.hpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#include "Logger.hpp"
+
+#include <stdexcept>
+
+/**
+ * @brief Name of the application entry function, replacing \c main().
+ */
+int aliceVision_main(int argc, char* argv[]);
+
+// Inline to ensure the right main() is used (libf2c also has one...)
```
Comment (fabiencastan) [inline src/aliceVision/system/main.hpp:28]: I would suggest a more explicit comment, like:
```
// Implementation of the unique main entry point.
// This method will call aliceVision_main and catch the exceptions to
// log the error message and avoid seg fault on exception on windows.
```

**Comment by fabiencastan** [comment]
It would be good to apply it on all `main_*.cpp` and not only `software/pipeline`.

File: src/aliceVision/system/main.hpp (hunk 1 above)
Comment (res2k) [reply src/aliceVision/system/main.hpp]: Ok, I expanded the comment.

**Comment by res2k** [comment]
> It would be good to apply it on all `main_*.cpp` and not only `software/pipeline`.

Yes. I would have done that as a follow-up PR, but I can add it to this one as well.

**Pull request merged**